pkg/document/json/rga_tree_split.go

```diff
@@ -492,3 +524,45 @@ func (s *RGATreeSplit) AnnotatedString() string {
 
 	return strings.Join(result, "")
 }
+
+// removedNodesLen returns length of removed nodes
+func (s *RGATreeSplit) removedNodesLen() int {
+	return len(s.removedNodeMap)
+}
+
+// cleanupRemovedNodes cleans up nodes that have been removed.
+// The cleaned nodes are subject to garbage collector collection.
+func (s *RGATreeSplit) cleanupRemovedNodes(ticket *time.Ticket) int {
+	count := 0
+	for _, node := range s.removedNodeMap {
+		if node.removedAt != nil && ticket.Compare(node.removedAt) >= 0 {
+			s.treeByIndex.Delete(node.indexNode)
+			s.purge(node)
+			s.treeByID.Remove(node.id)
+			delete(s.removedNodeMap, removedNodeMapKey(node.createdAt().Key()))
+			count++
+		}
+	}
+
+	return count
+}
+
+// purge removes the node passed as a parameter from RGATreeSplit.
+func (s *RGATreeSplit) purge(node *RGATreeSplitNode) {
+	if node.prev != nil {
+		node.prev.next = node.next
+		if node.next != nil {
+			node.next.prev = node.prev
+		}
```

Comment:
@dc7303 I have a question. is the purpose of the purge function to remove a specific node from theLinkedList?

Reply:
Thanks for the review! 👍 

When the text is modified, it does not remove the node for the deleted text, just tombstone marking. (Tombstone is `removedAt`)
This means that there are persistent nodes for deleted text, which can be a waste of memory.
We need to remove the tombstone marked nodes after synchronizing with other clients to fix this problem.

When executing `garbage collection for large size of text garbage test`, try to modify `RGATreeSplit.cleanupRemovedNodes` as shown below. Then you can see that the size of the object before and after the modification is different.

```go
// rga_tree_split.go
func (s *RGATreeSplit) cleanupRemovedNodes(ticket *time.Ticket) int {
	count := 0
	for _, node := range s.removedNodeMap {
		if node.removedAt != nil && ticket.Compare(node.removedAt) >= 0 {
			s.treeByIndex.Delete(node.indexNode)
			// s.purge(node)
			s.treeByID.Remove(node.id)
			delete(s.removedNodeMap, removedNodeMapKey(node.createdAt().Key()))
			count++
		}
	}

	return count
}
```

In addition, you can check the state of the tree through the `Text.AnnotatedString()` method within `doc.Update()`.

```go
	t.Run("garbage collection for text test", func(t *testing.T) {
		
		...

		// Example
		err = doc.Update(func(root *proxy.ObjectProxy) error {
			text := root.GetText("k1")
			fmt.Println(text.AnnotatedString())
			return nil
		}, "edit text k1")
	}

// Before s.purge(node) comment processing
//     [0:0:00:0 ][2:1:00:0 Hi][1:2:00:5  ][2:2:00:0 j][2:3:00:0 ane]
// After s.purge(node) comment processing
//     [2:1:00:0 Hi]{1:2:00:0 Hello}[1:2:00:5  ][2:2:00:0 j][2:3:00:0 ane]{1:3:00:0 m}{1:3:00:1 ario}{1:2:00:6 world}
```
---

I found out that the `garbage collection for large size of text garbage test` test was wrong thanks to your comments. Thanks to you, I modified this test. Thank you so much! 😄